src/main/scala/firrtl/transforms/CSESubAccesses.scala

```diff
@@ -21,7 +21,9 @@ object CSESubAccesses {
     val acc = new mutable.ListBuffer[(SubAccess, Info)]
     def onExpr(outer: Statement)(expr: Expression): Unit = {
       // Need postorder because we want to visit inner SubAccesses first
-      expr.foreach(onExpr(outer))
+      // Stop recursing on any non-Source because flips can make the SubAccess a Source despite the
+      //   overall Expression being a Sink
+      if (flow(expr) == SourceFlow) expr.foreach(onExpr(outer))
```

Comment:
To check my understanding... the implication of this is that we won't CSE _all_ subaccesses, and punt on ones which have weird internal flippedness? 

I think this is totally fair.

Reply:
This comment is really talking about how an Expression that is a Sink may have a SubAccess that is a SourceFlow, eg. from the added tests:
```
input out : { flip foo : UInt<8> }[2]
...
out[idx].foo <= in.foo
```
`out[idx].foo` is a sink, but the `out[idx]` SubAccess is SourceFlow, the `flip` on `foo` is what makes the larger Expression a sink.

Note that this case probably should not ever be handled by this transform, or at least, it would require a lot more sophistication to handle write ports

That being said, this check also stops handling any SubAccesses of bidirectional Vecs. Technically, this transform *could* handle them if running after `ExpandConnects`, but it's a little more complicated and I didn't feel like dealing with it, from the other test I added in this PR:
```
output out : { flip foo : UInt<8>, bar : UInt<8> }[4]
...
out[idx] <= in"
```